metricbeat/module/prometheus/query/data.go

```diff
@@ -0,0 +1,137 @@
+	ResultType string        `json:"resultType"`
+	Results    []interface{} `json:"result"`
+}
+
+// MapResponse is for "vector", "matrix" type from Prometheus Query API Request
+type MapResponse struct {
+	Status string  `json:"status"`
+	Data   mapData `json:"data"`
+}
+type mapData struct {
+	ResultType string      `json:"resultType"`
+	Results    []mapResult `json:"result"`
+}
+type mapResult struct {
+	Metric  map[string]string `json:"metric"`
+	Vector  []interface{}     `json:"value"`
+	Vectors [][]interface{}   `json:"values"`
+}
+
+func parseResponse(body []byte, pathConfig QueryConfig) ([]mb.Event, error) {
+	var events []mb.Event
+	converted, resultType, err := convertJSONToStruct(body)
+	if err != nil {
+		return events, err
+	}
+	switch resultType {
+	case "scalar", "string":
+		res := converted.(ArrayResponse)
+		events = append(events, mb.Event{
+			Timestamp: getTimestamp(res.Data.Results[0].(float64)),
+			MetricSetFields: common.MapStr{
+				"dataType":           resultType,
+				pathConfig.QueryName: convertToNumeric(res.Data.Results[1].(string)),
+			},
+		})
+	case "vector":
+		res := converted.(MapResponse)
+		for _, result := range res.Data.Results {
+			events = append(events, mb.Event{
+				Timestamp:    getTimestamp(result.Vector[0].(float64)),
+				ModuleFields: common.MapStr{"labels": result.Metric},
+				MetricSetFields: common.MapStr{
+					"dataType":           resultType,
+					pathConfig.QueryName: convertToNumeric(result.Vector[1].(string)),
+				},
+			})
+		}
+	case "matrix":
+		res := converted.(MapResponse)
+		for _, result := range res.Data.Results {
+			for _, vector := range result.Vectors {
+				events = append(events, mb.Event{
+					Timestamp:    getTimestamp(vector[0].(float64)),
+					ModuleFields: common.MapStr{"labels": result.Metric},
+					MetricSetFields: common.MapStr{
+						"dataType":           resultType,
+						pathConfig.QueryName: convertToNumeric(vector[1].(string)),
+					},
+				})
+			}
+		}
+	default:
+		return events, errors.New("Unknown resultType " + resultType)
+	}
+	return events, nil
+}
+
+func convertJSONToStruct(body []byte) (interface{}, string, error) {
+	arrayBody := ArrayResponse{}
+	if err := json.Unmarshal(body, &arrayBody); err != nil {
+		return nil, "", errors.Wrap(err, "Failed to parse api response")
+	}
+	if arrayBody.Status == "error" {
+		return nil, "", errors.Errorf("Failed to query")
+	}
+
+	if arrayBody.Data.ResultType == "vector" || arrayBody.Data.ResultType == "matrix" {
+		mapBody := MapResponse{}
+		if err := json.Unmarshal(body, &mapBody); err != nil {
+			return nil, arrayBody.Data.ResultType, errors.Wrap(err, "Failed to parse api response")
+		}
+		return mapBody, mapBody.Data.ResultType, nil
+	}
+	return arrayBody, arrayBody.Data.ResultType, nil
+}
+
+func getTimestamp(num float64) time.Time {
+	sec := int64(num)
+	ns := int64((num - float64(sec)) * 1000)
+	return time.Unix(sec, ns)
+}
+
+func convertToNumeric(str string) interface{} {
```

Comment:
Avoid code that returns `interface{}`

Reply:
Why those conversions if a `str` is also a valid value at the end?

Reply:
From what I can see, "values" of the response are always strings. Hence you want to try convert that strings into integers or floats otherwise to leave them strings. For instance:

"5" should be converted to 5, however "stable" should remain string.

A better name of this function could be `attemptConvertToNumeric`.